File: go/testutils/http_client.go
package testutils

import (
	"io"
	"net/http"
	"net/url"
)

type HttpClient struct {
	Res http.Response
}

func NewMockHttpClient(ress ...http.Response) *HttpClient {
	res := http.Response{}
	if len(ress) > 0 {
		res = ress[0]
	}
	if res.Header == nil {
		res.Header = http.Header{}
	}
	if res.Body == nil {
		res.Body = http.NoBody
	}

	return &HttpClient{Res: res}
}

func (*HttpClient) CloseIdleConnections() {

}
func (hc *HttpClient) Do(req *http.Request) (*http.Response, error) {
	res := hc.Res
	for k, v := range req.Header {
		for _, vv := range v {
			res.Header.Add(k, vv)
		}
	}
	if res.StatusCode == 0 {
		res.StatusCode = http.StatusOK
	}
	return &res, nil
}
func (*HttpClient) Get(url string) (resp *http.Response, err error) {
	return nil, nil
}
func (*HttpClient) Head(url string) (resp *http.Response, err error) {
	return nil, nil
}
func (*HttpClient) Post(url string, contentType string, body io.Reader) (resp *http.Response, err error) {
	return nil, nil
}
func (*HttpClient) PostForm(url string, data url.Values) (resp *http.Response, err error) {
	return nil, nil
}

File: ts/src/testutils/http_client.ts
import { HttpRequest } from "../types/http_request.js";
import { HttpClient } from "../types/http_client.js";
import { HttpResponse } from "../types/http_response.js";
import { HttpHeader } from "../types/http_header.js";
import { HttpURL } from "../types/http_url.js";

class httpClient implements HttpClient {
  readonly res: HttpResponse;
  req?: HttpRequest;

  constructor(res: HttpResponse) {
    this.res = res;
  }

  Do(req: HttpRequest): Promise<HttpResponse> {
    this.req = req;
    return Promise.resolve(this.res);
  }

  Get(url: string): Promise<HttpResponse> {
    return this.Do({
      Header: HttpHeader.from(),
      URL: HttpURL.parse(url).unwrap(),
      Method: "GET",
    });
  }

  Post(url: string, contentType: string, body: ReadableStream<Uint8Array>): Promise<HttpResponse> {
    return this.Do({
      Header: HttpHeader.from({
        "Content-Type": contentType,
      }),
      URL: HttpURL.parse(url).unwrap(),
      Method: "POST",
      Body: body,
    });
  }
  CloseIdleConnections(): void {
    // pass
  }
}

export function NewMockHttpClient(res: HttpResponse): HttpClient {
  return new httpClient(res);
}
